filc/retrieval.go

```diff
@@ -0,0 +1,407 @@
+	ipld "github.com/ipfs/go-ipld-format"
+	"github.com/ipfs/go-merkledag"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	"golang.org/x/xerrors"
+)
+
+type RetrievalCandidate struct {
+	Miner   address.Address
+	RootCid cid.Cid
+	DealID  uint
+}
+
+type CandidateSelectionConfig struct {
+	// Whether retrieval over IPFS is preferred if available
+	tryIPFS bool
+
+	// If true, candidates will be tried in the order they're passed in
+	// unchanged (and all other sorting-related options will be ignored)
+	noSort bool
+}
+
+type RetrievalResults struct {
+}
+
+func (node *Node) GetRetrievalCandidates(endpoint string, c cid.Cid) ([]RetrievalCandidate, error) {
+
+	endpointURL, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, xerrors.Errorf("endpoint %s is not a valid url", endpoint)
+	}
+	endpointURL.Path = path.Join(endpointURL.Path, c.String())
+
+	resp, err := http.Get(endpointURL.String())
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("http request to endpoint %s got status %v", endpointURL, resp.StatusCode)
+	}
+
+	var res []RetrievalCandidate
+
+	if err := json.NewDecoder(resp.Body).Decode(&res); err != nil {
+		return nil, xerrors.Errorf("could not unmarshal http response for cid %s", c)
+	}
+
+	return res, nil
+}
+
+type RetrievalStats interface {
+	GetByteSize() uint64
+	GetDuration() time.Duration
+	GetAverageBytesPerSecond() uint64
+}
+
+type FILRetrievalStats struct {
+	filclient.RetrievalStats
+}
+
+func (stats *FILRetrievalStats) GetByteSize() uint64 {
+	return stats.Size
+}
+
+func (stats *FILRetrievalStats) GetDuration() time.Duration {
+	return stats.Duration
+}
+
+func (stats *FILRetrievalStats) GetAverageBytesPerSecond() uint64 {
+	return stats.AverageSpeed
+}
+
+type IPFSRetrievalStats struct {
+	ByteSize uint64
+	Duration time.Duration
+}
+
+func (stats *IPFSRetrievalStats) GetByteSize() uint64 {
+	return stats.ByteSize
+}
+
+func (stats *IPFSRetrievalStats) GetDuration() time.Duration {
+	return stats.Duration
+}
+
+func (stats *IPFSRetrievalStats) GetAverageBytesPerSecond() uint64 {
+	return uint64(float64(stats.ByteSize) / stats.Duration.Seconds())
+}
+
+func (node *Node) RetrieveFromBestCandidate(
+	ctx context.Context,
+	fc *filclient.FilClient,
+	c cid.Cid,
+	candidates []RetrievalCandidate,
+	cfg CandidateSelectionConfig,
+) (RetrievalStats, error) {
+	// Try IPFS first, if requested
+	if cfg.tryIPFS {
+		log.Info("Searching IPFS for CID...")
+
+		log.Info("Bootstrapping DHT...")
+		bootstrapPeers := dht.GetDefaultBootstrapPeerAddrInfos()
+
+		dht, err := dht.New(
+			ctx,
+			node.Host,
+			dht.Mode(dht.ModeClient),
+		)
+		if err != nil {
+			// TODO: don't die here
+			return nil, err
+		}
+
+		// Connect to IPFS peers in parallel
+
+		var peersConnected sync.WaitGroup
+		var peersLk sync.Mutex
+		peersConnectedCount := 0
+
+		peersConnected.Add(len(bootstrapPeers))
+
+		for _, ai := range bootstrapPeers {
+			ai := ai
```

Comment:
you can avoid having to do a local bind copy like this if you just make the anonymous function accept the type.